src/plugin.ts

```diff
@@ -21,31 +33,37 @@ type UpdateHeads = {
 
 export const effectType = StateEffect.define<UpdateHeads>({})
 
-export function updateHeads(newHeads: Heads): StateEffect<UpdateHeads> {
-  return effectType.of({ newHeads })
-}
+export const updateHeads = (newHeads: Heads): StateEffect<UpdateHeads> =>
+  effectType.of({ newHeads })
 
-export function getLastHeads(state: EditorState, field: Field): Heads {
-  return state.field(field).lastHeads
-}
+export const getLastHeads = (state: EditorState, field: Field): Heads =>
+  state.field(field).lastHeads
 
-export function getPath(state: EditorState, field: Field): Prop[] {
-  return state.field(field).path
-}
+export const getPath = (state: EditorState, field: Field): Prop[] =>
+  state.field(field).path
 
 export type Field = StateField<Value>
 
-export function plugin<T>(doc: Doc<T>, path: Prop[]): StateField<Value> {
-  return StateField.define({
-    create() {
-      return {
-        lastHeads: automerge.getHeads(doc),
-        unreconciledTransactions: [],
-        path: path.slice(),
-      }
-    },
-    update(value: Value, tr: Transaction) {
-      const result = {
+const semaphoreFacet = Facet.define<PatchSemaphore, PatchSemaphore>({
+  combine: values => values.at(-1)!, // Take last.
+})
+
+export type AutomergePlugin = {
+  extension: Extension
+}
+
+export const automergePlugin = (
+  handle: IDocHandle,
+  path: Prop[]
+): AutomergePlugin => {
+  const stateField: StateField<Value> = StateField.define({
+    create: () => ({
+      lastHeads: automerge.getHeads(handle.docSync()!),
```

Comment:
This feels like a bit of a footgun. If a user passes a handle which is not ready yet then this will throw an error. I think we have three options:

1. Add a "loading" state to the plugin and then attach a `then` handle to `handle.doc()` which transitions to a "ready" state. In the "loading" state the plugin would not allow input.
2. Make `automergePlugin` `async` and wait for `handle.doc().await`
3. Document in big letters that you must make sure the handle is ready before passing it to the plugin.

My preference would be for `2`. Thoughts?

Reply:
Good catch. 

Having a loading state seems to be out of scope for the editor plugin. Developers would want to customize the look of the loading state anyway or integrate with their UI library like react suspense.

I'd be hesitant to make the plugin  async as there's no precedent for it. Main alternative - yjs, has a sync plugin. It also makes using wrappers that take extensions in props like https://uiwjs.github.io/react-codemirror/ more difficult.

I think my preference would be on option #3 with a clear error message that directs developers to implement a loading state for the editor. This seems to be the most flexible option in terms of API design.

Reply:
Good point on the extension props. I agree with your argument, #3 is probably the best we can do.

Reply:
seconding what's being said, as a developer I think not implementing the loading state in the plugin is the right way to go - off the top of my head I can think of two ways one might want to integrate it, either not showing the editor at all while loading or having a CM `Compartment` that is initially empty but is updated to contain the plugin once the handle is ready.

Reply:
@dmaretskyi is adding an error message when passed a `DocHandle` with an `undefined` `doc` something you have capacity for? I think it would be really useful to merge this PR so would love to get that in there.